crates/utils/src/utils/markdown/mod.rs

```diff
@@ -0,0 +1,217 @@
+use crate::settings::SETTINGS;
+use markdown_it::{plugins::cmark::inline::image::Image, MarkdownIt};
+use once_cell::sync::Lazy;
+use url::Url;
+use urlencoding::encode;
+
+mod link_rule;
+mod spoiler_rule;
+
+static MARKDOWN_PARSER: Lazy<MarkdownIt> = Lazy::new(|| {
+  let mut parser = MarkdownIt::new();
+  markdown_it::plugins::cmark::add(&mut parser);
+  markdown_it::plugins::extra::add(&mut parser);
+  spoiler_rule::add(&mut parser);
+  link_rule::add(&mut parser);
+
+  parser
+});
+
+/// Replace special HTML characters in API parameters to prevent XSS attacks.
+///
+/// Taken from https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md#output-encoding-for-html-contexts
+///
+/// `>` is left in place because it is interpreted as markdown quote.
+pub fn sanitize_html(text: &str) -> String {
+  text
+    .replace('&', "&amp;")
+    .replace('<', "&lt;")
+    .replace('\"', "&quot;")
+    .replace('\'', "&#x27;")
+}
+
+pub fn markdown_to_html(text: &str) -> String {
+  MARKDOWN_PARSER.parse(text).xrender()
+}
+
+/// Rewrites all links to remote domains in markdown, so they go through `/api/v3/image_proxy`.
+pub fn markdown_rewrite_image_links(mut src: String) -> String {
+  let ast = MARKDOWN_PARSER.parse(&src);
+  let mut links = vec![];
+
+  // Walk the syntax tree to find positions of image links
+  ast.walk(|node, _depth| {
+    if let Some(image) = node.cast::<Image>() {
+      let node_offsets = node.srcmap.expect("srcmap is none").get_byte_offsets();
+      let start_offset = node_offsets.1 - image.url.len() - 1;
+      let end_offset = node_offsets.1 - 1;
+
+      links.push((start_offset, end_offset));
+    }
+  });
+
+  // Go through the collected links
+  while let Some((start, end)) = links.pop() {
+    let url = &src[start..end];
+    match Url::parse(url) {
+      Ok(parsed) => {
+        // If link points to remote domain, replace with proxied link
+        if parsed.domain() != Some(&SETTINGS.hostname) {
+          let proxied = format!(
+            "{}/api/v3/image_proxy?url={}",
+            SETTINGS.get_protocol_and_hostname(),
+            encode(&url)
+          );
```

Comment:
Here we could also write a relative url as in `/api/v3/image_proxy...` which would be a bit simpler and shorter. It would also make domain changes a bit easier. Or is there some argument in favor of full urls with domain?

Reply:
Yeah IMO its not worth it, there's already a lot of things that need to be rewritten for domain name changes anyway.